Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -41,9 +90,108 @@ public struct DiagnosticsFormatter {
     return formatter.annotatedSource(tree: tree, diags: diags)
   }
 
+  /// Colorize the given source line by applying highlights from diagnostics.
+  private func colorizeSourceLine<SyntaxType: SyntaxProtocol>(
+    _ annotatedLine: AnnotatedSourceLine,
+    lineNumber: Int,
+    tree: SyntaxType,
+    sourceLocationConverter slc: SourceLocationConverter
+  ) -> String {
+    guard colorize, !annotatedLine.diagnostics.isEmpty else {
+      return annotatedLine.sourceString
+    }
+
+    // Compute the set of highlight ranges that land on this line. These
+    // are column ranges, sorted in order of increasing starting column, and
+    // with overlapping ranges merged.
+    let highlightRanges: [Range<Int>] = annotatedLine.diagnostics.map {
+      $0.highlights
+    }.joined().compactMap { (highlight) -> Range<Int>? in
+      if highlight.root != Syntax(tree) {
+        return nil
+      }
+
+      let startLoc = highlight.startLocation(converter: slc, afterLeadingTrivia: true);
+      guard let startLine = startLoc.line else {
+        return nil
+      }
+
+      // Find the starting column.
+      let startColumn: Int
+      if startLine < lineNumber {
+        startColumn = 1
+      } else if startLine == lineNumber, let column = startLoc.column {
+        startColumn = column
+      } else {
+        return nil
+      }
+
+      // Find the ending column.
+      let endLoc = highlight.endLocation(converter: slc, afterTrailingTrivia: false)
+      guard let endLine = endLoc.line else {
+        return nil
+      }
+
+      let endColumn: Int
+      if endLine > lineNumber {
+        endColumn = annotatedLine.sourceString.count
+      } else if endLine == lineNumber, let column = endLoc.column {
+        endColumn = column
+      } else {
+        return nil
+      }
+
+      if startColumn == endColumn {
+        return nil
+      }
+
+      return startColumn..<endColumn
+    }.sorted { (lhs, rhs) in
+      lhs.lowerBound < rhs.lowerBound
+    }.mergingOverlappingRanges()
+
+    // Map the column ranges into index ranges within the source string itself.
+    let sourceString = annotatedLine.sourceString
+    let highlightIndexRanges: [Range<String.Index>] = highlightRanges.map { highlightRange in
+      let startIndex = sourceString.index(sourceString.startIndex, offsetBy: highlightRange.lowerBound - 1)
```

Comment:
I’m not 100% sure but I don’t think this is Unicode correct. The columns are about UTF-8 byte offsets whereas `String.Index` is about Unicode characters/grapheme clusters/something along those lines. IIRC there’s a way to convert between the two but I don’t remember what it is. Adding a few tests with emojis would certainly be good.

Reply:
Oh, I think you're right. I should be computing indices via the UTF-8 view of the source string.

Reply:
Yes, I got this fixed up in https://github.com/apple/swift-syntax/pull/1365

Reply:
Thanks!